File: cpp/LeetCode/No746.cpp
/*
You are given an integer array cost where cost[i] is the cost of ith step on a staircase. 
Once you pay the cost, you can either climb one or two steps.

You can either start from the step with index 0, or the step with index 1.

Return the minimum cost to reach the top of the floor.
*/
#include <iostream>
#include <vector>
#include <cmath>
#include "Debug\VectorDebug.h"

using namespace std;

class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();
        int first = cost[0];
        int second = cost[1];
        int third;
        for(int i=2;i<n;i++){
            third = min(first, second) + cost[i];
            first = second;
            second = third;
        }
        return min(first, second);
    }
};

int main(){
    Solution solution;

    vector<int> cost = {1,100,1,1,1,100,1,1,100,1};
    cout << solution.minCostClimbingStairs(cost) << endl;
    return 0;
}

File: python/LeetCode Quiz/No746.py
from typing import List

class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        mem = dict()
        def climb(n: int) -> int:
            if n < 2:
                return 0
            if n not in mem:
                mem[n] = min(climb(n-1) + cost[n-1], climb(n-2)+cost[n-2])
            return mem[n]
        return climb(len(cost))
    
solution = Solution()
cost = [10,15,20]
print(solution.minCostClimbingStairs(cost))